crates/adapters/src/transport/clock.rs

```diff
@@ -0,0 +1,450 @@
+
+        let config_clone = config.clone();
+        TOKIO.spawn(async {
+            let _ = Self::worker_task(config_clone, parser, consumer, receiver).await;
+        });
+
+        Self { sender }
+    }
+
+    /// Current timestamp in milliseconds, rounded to `clock_resolution_ms`.
+    fn current_time(config: &ClockConfig) -> u64 {
+        let now = SystemTime::now();
+        let now_millis = now
+            .duration_since(UNIX_EPOCH)
+            .unwrap_or_default()
+            .as_millis() as u64;
+        let clock_resolution_ms = config.clock_resolution_ms();
+        now_millis - now_millis % clock_resolution_ms
+    }
+
+    /// Time when we want to trigger the next clock tick, assuming we triggered one at `previous_tick`.
+    ///
+    /// Returns time as `Instant`, so it can be used with `sleep_until`.
+    fn next_tick_time(config: &ClockConfig, previous_tick: &u64) -> Instant {
+        let next_tick = previous_tick + config.clock_resolution_ms();
+
+        let target_time = UNIX_EPOCH + Duration::from_millis(next_tick);
+        let now = SystemTime::now();
+        let duration_until = target_time.duration_since(now).unwrap_or(Duration::ZERO);
+
+        Instant::now() + duration_until
+    }
+
+    /// Convert timestamp to a JSON object that can be fed to the parser.
+    fn timestamp_to_record(ts_millis: u64) -> String {
+        format!("{{\"now\":{ts_millis}}}")
+    }
+
+    async fn worker_task(
+        config: Arc<ClockConfig>,
+        mut parser: Box<dyn Parser>,
+        consumer: Box<dyn InputConsumer>,
+        mut receiver: UnboundedReceiver<InputReaderCommand>,
+    ) {
+        let mut next_tick: Option<Instant> = None;
+        loop {
+            select! {
+                // When the next clock tick is scheduled, wakeup at `next_tick` time.
+                _ = sleep_until(next_tick.unwrap_or_else(Instant::now)), if next_tick.is_some() => {
+                    next_tick = None;
+                    consumer.request_step();
+                }
+                message = receiver.recv() => match message {
+                    None => {
+                        // channel closed
+                        break;
+                    }
+                    // Nothing to do for Seek: the connector simply starts from the current timestamp.
+                    Some(InputReaderCommand::Seek(..)) => {}
+                    Some(InputReaderCommand::Replay { data, .. }) => {
+                        // Parse serialized timestamp, push it to the circuit.
+                        let Some(ts_millis) = (match data {
+                            RmpValue::Integer(int) => int.as_u64(),
+                            _ => None,
+                        }) else {
+                            consumer.error(true, anyhow!("Invalid timestamp in replay log: {data:?}"));
+                            continue;
+                        };
+                        consumer.buffered(1, 8);
```

Comment:
`8` is easy to misread (it looks a lot like zero), maybe I would write `std::mem::size_of::<u64>()` to make it obvious what's being measured here.